### PR TITLE
feat: add window aggregation data operation package (1/9)

### PR DESCRIPTION
**Pull request by TKaltofen**
## Summary

- Implements `WindowAggregationFeatureGroup` with `{col}__{agg}_groupby` pattern (broadcast, row-preserving)
- 4 framework implementations: PyArrow (reference), SQL/SQLite, Polars Lazy, Pandas
- Shared data operations infrastructure: `base.py` (config keys, null policy, type-check helpers), `DataOperationsTestDataCreator` (12-row test dataset)
- 12 supported aggregation operations: sum, avg, count, min, max, std, var, median, mode, nunique, first, last
- SQL implementation follows PEP 249 with parameterized queries and validated identifiers

## Test plan

- [x] 34 base class tests (pattern matching, config validation, parsing)
- [x] 22 PyArrow tests (all operations, null handling, edge cases, multi-key partition)
- [x] 11 SQL tests (native SQL window functions, null handling, row preservation)
- [x] 10 Polars Lazy tests (basic operations, null handling, row preservation)
- [x] 10 Pandas tests (basic operations, null handling with dropna=False, row preservation)
- [x] 5 cross-framework comparison tests (all 4 frameworks match PyArrow reference)
- [x] 9 integration tests (mloda.run_all pipeline, PluginCollector discovery)
- [x] tox green: pytest (119 passed, 3 skipped) + ruff + mypy + bandit
- [x] Security review: SQL injection mitigated, identifier validation added

**Ticket:** `tickets/229-data-operation-feature-groups.md`
**Task:** `agents/tasks/039-data-ops-window-aggregation.md`

### CODE REVIEW

File: mloda/community/feature_groups/data_operations/base.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        ),
+        ColumnTypeCategory.STRING: lambda t: pa.types.is_string(t) or pa.types.is_large_string(t),
+        ColumnTypeCategory.DATETIME: lambda t: pa.types.is_timestamp(t) or pa.types.is_date(t),
+        ColumnTypeCategory.ANY: lambda t: True,
+    }
+
+
+def _pandas_type_checkers() -> dict[ColumnTypeCategory, Any]:
+    """Return pandas dtype-checking predicates, imported lazily."""
+    import pandas as pd
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/base.py]: why is pandas code in base?

File: mloda/community/feature_groups/data_operations/base.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+    def is_datetime(dtype: Any) -> bool:
+        return bool(dtype == pl.Datetime or dtype == pl.Date)
+
+    return {
+        ColumnTypeCategory.NUMERIC: is_numeric,
+        ColumnTypeCategory.STRING: is_string,
+        ColumnTypeCategory.DATETIME: is_datetime,
+        ColumnTypeCategory.ANY: lambda dtype: True,
+    }
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/base.py]: polars issue

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/tests/test_duckdb.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+def _extract_column(result: DuckdbRelation, column_name: str) -> list[Any]:
+    """Extract a column from a DuckdbRelation as a Python list."""
+    return list(result.to_arrow_table().column(column_name).to_pylist())
+
+
+class TestDuckdbBasicAggregations:
+    """Test sum, avg, count, min, max with pre-computed expected values."""
+
+    def test_sum_groupby_region(self, sample_data: DuckdbRelation) -> None:
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/tests/test_duckdb.py]: why not have under testing/feature_groups/data_operations/row_preserving/ test utilities for window aggregation? It seems there is quite some overlap

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/sqlite_window_aggregation.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+_SQLITE_AGG_FUNCS: dict[str, str] = {
+    "sum": "SUM",
+    "avg": "AVG",
+    "count": "COUNT",
+    "min": "MIN",
+    "max": "MAX",
+}
+
+# Aggregation types that require a Python fallback (no native SQLite window support).
+_PYTHON_FALLBACK_AGGS = {"std", "var", "median", "mode", "nunique", "first", "last"}
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/sqlite_window_aggregation.py]: Is this used anywhere?

File: mloda/testing/pyproject.toml [hunk 5]
```diff
@@ -23,4 +23,4 @@ Repository = "https://github.com/mloda-ai/mloda-registry"
 
 [tool.setuptools]
 package-dir = {"" = "../.."}
-packages = ["mloda.testing"]
+packages = ["mloda.testing", "mloda.testing.data_creator", "mloda.testing.feature_groups", "mloda.testing.feature_groups.data_operations", "mloda.testing.feature_groups.data_operations.row_preserving"]
```
Comment (TomKaltofen) [inline mloda/testing/pyproject.toml]: Do we really need mloda.testing.feature_groups and mloda.testing.feature_groups.data_operations.row_preserving?

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/base.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+    # -- Shared aggregation helpers --
+
+    @classmethod
+    def _aggregate(cls, values: list[Any], agg_type: str) -> Any:
+        """Compute a single aggregate over a list of values (may contain None)."""
+        non_null = [v for v in values if v is not None]
+
+        if not non_null:
+            return None
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/base.py]: Is this the right behaviour? be critical

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/sqlite_window_aggregation.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        data: SqliteRelation,
+        feature_name: str,
+        source_col: str,
+        partition_by: list[str],
+        agg_type: str,
+    ) -> SqliteRelation:
+        """Delegate to SQL path for native aggs, Python fallback otherwise."""
+        if agg_type in _SQLITE_AGG_FUNCS:
+            return cls._compute_window_sql(data, feature_name, source_col, partition_by, agg_type)
+        return cls._compute_window_python(data, feature_name, source_col, partition_by, agg_type)
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/sqlite_window_aggregation.py]: why do we have compute window python? If it is not supported, it is simply not supported. We should not have it.

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/pyarrow_window_aggregation.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        for key, values in groups.items():
+            agg_results[key] = cls._aggregate(values, agg_type)
+
+        # Broadcast back to every row
+        result_values = [agg_results[keys[i]] for i in range(num_rows)]
+
+        new_col = pa.array(result_values)
+        return table.append_column(feature_name, new_col)
+
+    # -- Aggregation helpers (Python-based, used by this framework only) --
```
Comment (TomKaltofen) [inline mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/pyarrow_window_aggregation.py]: Can you explain this more why?

File: mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/base.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        "nunique": "Count of unique values",
+        "first": "First value in partition",
+        "last": "Last value in partition",
+    }
+
+    # partition_by is intentionally excluded from PROPERTY_MAPPING because the
+    # mixin's _process_found_property_value cannot handle list-valued options.
+    # Validation happens in match_feature_group_criteria override instead.
+    # See https://github.com/mloda-ai/mloda/issues/228
+    PROPERTY_MAPPING = {
```
[the remark this answers is not in the document]
Comment (TomKaltofen) [reply mloda/community/feature_groups/data_operations/row_preserving/window_aggregation/base.py]: It is released as 0.5.5
